src/NumericalAlgorithms/LinearSolver/Gmres.cpp

```diff
@@ -0,0 +1,31 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "NumericalAlgorithms/LinearSolver/Gmres.hpp"
+
+#include <tuple>
+
+#include "DataStructures/DenseMatrix.hpp"
+#include "DataStructures/DenseVector.hpp"
+
+namespace LinearSolver {
+namespace gmres_detail {
+
+std::pair<DenseVector<double>, double> minimal_residual(
+    const DenseMatrix<double>& orthogonalization_history,
+    const double initial_residual_magnitude) noexcept {
+  // Perform a QR decomposition of the Hessenberg matrix that was built during
```

Comment:
This is performing the part of the algorithm that is an explicit linear solve by QR decomposition, right? There's almost certainly a LAPACK version that does this. I don't think they have a specialization for upper Hessenberg, but I'd expect the LAPACK version to perform better than manually handling the inverse yourself.
It might also be good to consider passing in a buffer to avoid the matrix allocations, but I'm less sure about that because the buffer would be changing significantly in size during the execution of the GMRES algorithm if I understand correctly.

Reply:
This function computes a linear least-squares problem to find the vector `y` that minimizes the function `||beta - Hy||`, where `beta` is the vector `(initial_residual_magnitude, 0, 0, ...)` of the same length as `y` and `H` is the `orthogonalization_history` Hessenberg matrix. I could probably replace the QR decomposition done through blaze by a call to [`sgels`](https://www.netlib.org/lapack/lug/node27.html). Do you think that would be worthwhile?

Reply:
Actually I think to make this function more efficient I should go another way: Since `H` only gains a non-zero column from one iteration to the next, and is Hessenberg, we can iteratively update the QR decomposition.

Reply:
Yeah, I would anticipate that the `sgels` call to save on memory traffic and take better advantage of the low-level optimizations that LAPACK performs -- ultimately I anticipate that the individual blaze calls are separately dispatching to LAPACK internally, but we should probably expect that LAPACK will do a better job of managing the collection of calls whenever it has a full algorithm available for the problem you're solving.

Reply:
Ah, yeah, okay -- if you can manage to avoid duplicating work on the inversion step, that would probably be better.